lib/sidekiq/job_util.rb

```diff
@@ -70,50 +70,41 @@ def normalized_hash(item_class)
 
     private
 
-    RECURSIVE_JSON_SAFE = {
-      Integer => ->(val) { true },
-      Float => ->(val) { true },
-      TrueClass => ->(val) { true },
-      FalseClass => ->(val) { true },
-      NilClass => ->(val) { true },
-      String => ->(val) { true },
+    RECURSIVE_JSON_UNSAFE = {
+      Integer => ->(val) {},
+      Float => ->(val) {},
+      TrueClass => ->(val) {},
+      FalseClass => ->(val) {},
+      NilClass => ->(val) {},
+      String => ->(val) {},
       Array => ->(val) {
-        val.all? { |e| RECURSIVE_JSON_SAFE[e.class].call(e) }
+        val.each do |e|
+          unsafe_item = RECURSIVE_JSON_UNSAFE[e.class].call(e)
+          return unsafe_item unless unsafe_item.nil?
+        end
+        nil
       },
       Hash => ->(val) {
-        val.all? { |k, v| String === k && RECURSIVE_JSON_SAFE[v.class].call(v) }
+        val.each do |k, v|
+          return k unless String === k
+
+          unsafe_item = RECURSIVE_JSON_UNSAFE[v.class].call(v)
+          return unsafe_item unless unsafe_item.nil?
+        end
+        nil
       }
     }
 
-    RECURSIVE_JSON_SAFE.default = ->(_val) { false }
-    RECURSIVE_JSON_SAFE.compare_by_identity
-    private_constant :RECURSIVE_JSON_SAFE
+    RECURSIVE_JSON_UNSAFE.default = ->(val) { val }
+    RECURSIVE_JSON_UNSAFE.compare_by_identity
+    private_constant :RECURSIVE_JSON_UNSAFE
 
     def json_safe?(item)
-      RECURSIVE_JSON_SAFE[item.class].call(item)
+      json_unsafe_item(item).nil?
```

Comment:
I think you're calling `json_unsafe_item` twice when the payload is unsafe? If you switch this method to `json_unsafe?` and return the value rather than using `.nil?` you can avoid the second call.

Reply:
Yes, thanks! Fixed.

Reply:
Eh, 2 commits, please squash when merging.